### PR TITLE
task week streamline proccess

### PR DESCRIPTION
**Pull request by benjaminmielke**
Created a Class that hold methods that can be used as a pipeline for each task week. Pipeline from import data to preprocess to model and evaluation. Hope this helps us with the data processing each week. I am no pro at python code structure, but figured this way has some flexibility. We can add other methods for pre processing and other models. Would love to hear if you think this type of structure makes sense, i am just not sure

### CODE REVIEW

File: scripts/challenge_week.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        return self.df_demand, self.df_pv, self.df_weather, self.df_submit
+
+    def merge_dataframes(self):
+        '''Creates two combined dataframes from raw task week data.
+
+        Returns:
+            df_train: dataframe containing all weather data, pv_power data, and demand data
+                      given for the task week. There will be missing weather data at this point,
+                      becuase the weather data is hourly.
+            df_taskweek: dataframe containing all the wather data that can be used for
```
Comment (rjbergerud) [inline scripts/challenge_week.py:51]: typo *weather

Comment (rjbergerud) [reply scripts/challenge_week.py:51]: *week

File: scripts/challenge_week.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+                         prediction of the wekk of interest.
+
+        '''
+        # Check 'datetime' columns are equal in demand and pv dfs, this must be true
+        assert self.df_demand['datetime'].equals(self.df_pv['datetime']), "Demand and PV train sets have different timestamps"
```
Comment (rjbergerud) [inline scripts/challenge_week.py:56]: Have you considered making the datetime column the index for the dataframe?  I can't think of an advantage to having the default integer index at the moment.

File: scripts/challenge_week.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        Args:
+            lst_cols(list): list of weather columns to impute
+            df: dataframe to which to pply the transformation
+        Returns:
+            df: transformed df
+        '''
+        for c in range(0, len(lst_cols)):
+            for i in range(0, len(df)-1):
+                if pd.isnull(df[lst_cols[c]].iloc[i]):
+                    df.loc[i, lst_cols[c]] = mean([df.loc[i-1, lst_cols[c]], df.loc[i+1, lst_cols[c]]])
```
Comment (rjbergerud) [inline scripts/challenge_week.py:101]: Could potentially get a key error here at i = 0, `df.loc[i-1, ...]`

File: scripts/challenge_week.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+            df[lst_cols[c]].iloc[len(df)-1] = df[lst_cols[c]].iloc[len(df)-2]
```
Comment (rjbergerud) [inline scripts/challenge_week.py:102]: Sorry, I'm lost as to what this line is doing.  Why rewrite the last value of each column with the one before it (if I am reading this correctly)?

File: scripts/challenge_week.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        '''Imputes the missing data for the weather paramters using the mean of the
+        timesteps before and after.
+
+        Args:
+            lst_cols(list): list of weather columns to impute
+            df: dataframe to which to pply the transformation
+        Returns:
+            df: transformed df
+        '''
+        for c in range(0, len(lst_cols)):
```
Comment (rjbergerud) [inline scripts/challenge_week.py:98]: Don't have to use it, but pandas does have some neat functions to facilitate this, `bfill` and `ffill`, and can be used in combination to get an average as per https://stackoverflow.com/questions/44102794/input-missed-values-with-mean-of-nearest-neighbors-in-column

**Review by rjbergerud — COMMENTED** [review]
Mostly small things, a few typos, etc., but do want to discuss choice of index before merging this pull request.  What are your thoughts on datetime index vs the integer index?  Using a datetime index and pandas `join` function on the index instead of `merge` on the datetime column just seems to me to make more sense, since it's defines the way the data is ordered and joined.  Also, pandas has a lot of built-in features for datetime indices https://towardsdatascience.com/basic-time-series-manipulation-with-pandas-4432afee64ea

File: scripts/challenge_week.py (hunk 4 above)
Comment (benjaminmielke) [reply scripts/challenge_week.py:102]: With the method I am using, if the last value in the column was NaN, it wouldn't be able to find the mean using the values immediately before and after, so I just assigned it the same value as the value immediately before it. Shouldnt be a huge deal considering the weather parameters usually wouldnt change too much in 30 minutes, but hopefully the Bfill and Ffill method could help with some of the short comings of this method.

File: scripts/challenge_week.py (hunk 3 above)
Comment (benjaminmielke) [reply scripts/challenge_week.py:101]: Yes that is a problem, the first value in the column should no be part of this line, as there is not a value before it to use for the mean.

File: scripts/challenge_week.py (hunk 5 above)
Comment (benjaminmielke) [reply scripts/challenge_week.py:98]: I was wondering, surely someone has a pandas method written up to do this better than what I have....i shall check this out

File: scripts/challenge_week.py (hunk 1 above)
Comment (benjaminmielke) [reply scripts/challenge_week.py:51]: gracias

File: scripts/challenge_week.py (hunk 2 above)
Comment (benjaminmielke) [reply scripts/challenge_week.py:56]: I am not exactly why I haven't kept the datetime as the index to be honest. I will probably change that to have it as the index

File: scripts/challenge_week.py (hunk 4 above)
Comment (rjbergerud) [reply scripts/challenge_week.py:102]: Ok, that makes sense, but isn't it outside the block that checks if that entry is NaN?

**Pull request merged**